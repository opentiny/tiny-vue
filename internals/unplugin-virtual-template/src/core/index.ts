export * from './transformVirtualTemplate'
export * from './transformConditionalTemplate'
export * from './transformVirtualTemplateForBuild'
