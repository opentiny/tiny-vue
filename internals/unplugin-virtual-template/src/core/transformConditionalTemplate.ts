import MagicString from 'magic-string'

export const transformConditionalTemplate = (code: string, id: string, env: any = {}) => {
  const source = new MagicString(code, {
    filename: id
  })

  source.replace(
    /^.*?#v-if(n?)def\s*(\S+).*[\r\n]{1,2}([\s\S]+?)\s*.*?#v-endif.*?$/gm,
    /**
     * 条件替换
     * @param _ 匹配的字符串
     * @param $1 是否为 not 模式
     * @param $2 条件
     * @param $3 code
     */
    (_, $1, $2, $3) => {
      const isNot = !!$1
      const isKeep = $2.split('||').some((v: string) => {
        let flag = false
        const [key, value] = v.split('=')
        if (value === undefined) {
          flag = !!env[key]
        } else {
          flag = String(env[key]) === value
        }
        flag = isNot ? !flag : flag
        return flag
      })
      return isKeep ? $3 : ''
    }
  )

  if (source.hasChanged()) {
    return {
      code: source.toString(),
      map: source.generateMap({
        source: id,
        file: `${id}.map`,
        includeContent: true
      })
    }
  }
}
