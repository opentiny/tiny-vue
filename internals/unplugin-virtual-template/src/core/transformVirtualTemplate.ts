export function transformVirtualTemplate(code: string) {
  return code.replace(/import\s+(\w+)\s+from\s+'virtual-template\?(.+)'/, (all, localName, url) => {
    const params = url
      .split('|')
      .map((item: string) => item.split('='))
      .reduce((result: Record<string, string>, [key, value]: [string, string]) => {
        result[key] = value || key
        return result
      }, {})

    const getTemplate = (params: Record<string, string>) => {
      return Object.entries(params)
        .map(([key, value]) => {
          return `  
  // #v-ifdef VITE_TINY_MODE=${key}
  result = import.meta.glob('./${value}.vue', { eager: true })
  // #v-endif
        `
        })
        .join('')
    }

    const result = `
const ${localName} = (mode) => {
  let result

  // #v-ifndef VITE_TINY_MODE
  result = import.meta.glob('./{pc,mobile,mobile-first}.vue', { eager: true })
  // #v-endif

  ${getTemplate(params)}

  const finalMode = ${JSON.stringify(params)}[mode?.value || mode] || '${params[Object.keys(params)[0]]}'

  return result[\`./\${finalMode}.vue\`].default
}
  `

    return result
  })
}
