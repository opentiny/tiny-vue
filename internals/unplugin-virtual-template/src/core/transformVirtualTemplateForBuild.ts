const toCamelCase = (str = '') => {
  const s =
    str &&
    str
      .match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
      .map((x: string) => x.slice(0, 1).toUpperCase() + x.slice(1).toLowerCase())
      .join('')
  return s.slice(0, 1).toUpperCase() + s.slice(1)
}

export function transformVirtualTemplateForBuild(code: string) {
  return code.replace(/import\s+(\w+)\s+from\s+'virtual-template\?(.+)'/, (all, localName, url) => {
    const params = url
      .split('|')
      .map((item: string) => item.split('='))
      .reduce((result: Record<string, string>, [key, value]: [string, string]) => {
        result[toCamelCase(`${key}-template`)] = value || key
        return result
      }, {})

    const getImports = (params: Record<string, string>) => {
      return Object.entries(params)
        .map(([key, value]) => {
          return `import ${key} from './${value}.vue'`
        })
        .join('\n')
    }

    const getTemplateFunction = (params: Record<string, string>) => {
      const templateArr = Object.entries(params)
      let defaultReturnStr = ''

      const templateStr = templateArr
        .map(([key, value], index) => {
          // 当模板个数等于1时，不需要默认return 语句
          if (!index && templateArr.length > 1) {
            defaultReturnStr = `return ${key}`
          }

          // 当模板数等于1则直接return该模板
          if (templateArr.length === 1) {
            return `return ${key}`
          } else {
            return `if ('${value}' === (tinyMode || mode)) {return ${key}}`
          }
        })
        .join('\n')

      return `${templateStr}${defaultReturnStr}`
    }

    const result = `
${getImports(params)}

const ${localName} = (mode) => {

  const tinyMode = typeof process === 'object' ? process.env?.TINY_MODE : null

  ${getTemplateFunction(params)}
}
  `
    return result
  })
}
