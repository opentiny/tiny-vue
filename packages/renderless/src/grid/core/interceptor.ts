/**
 * MIT License
 *
 * Copyright (c) 2019 Xu Liangzhan
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

import { toString, each, remove } from '../static'

const toType = (type) => toString(type).replace('_', '').toLowerCase()
const eventTypes = [
  'created',
  'mounted',
  'activated',
  'beforeDestroy',
  'destroyed',
  'event.clearActived',
  'event.clearFilter',
  'event.showMenu',
  'event.keydown'
].map(toType)
const _storeMap = {}

const Interceptor = {
  mixin(map) {
    each(map, (callback, type) => Interceptor.add(type, callback))
    return Interceptor
  },
  get(type) {
    return _storeMap[toType(type)] || []
  },
  add(type, callback) {
    const selfType = toType(type)

    if (callback && ~eventTypes.indexOf(selfType)) {
      _storeMap[selfType] = _storeMap[selfType] ? _storeMap[selfType] : []

      _storeMap[selfType].push(callback)
    }

    return Interceptor
  },
  delete(type, callback) {
    const eList = _storeMap[toType(type)]

    if (eList) {
      remove(eList, (cb) => cb === callback)
    }

    return Interceptor
  }
}

export default Interceptor
