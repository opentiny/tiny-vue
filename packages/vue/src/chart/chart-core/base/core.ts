import tips from './util/tips'
import merge from './util/merge'
import * as echarts from 'echarts'
import Theme from './feature/token'
import throttle from './util/throttle'
import axistip from './feature/axistip'
import { mergeExtend } from './util/merge'
import BaseChart from './components/BaseChart'
import readScreen from './feature/readScreen'
import MediaScreen from './feature/mediaScreen'

const SELF_CHART = [
  'FlowChart',
  'WaveChart',
  'RiverChart',
  'GanttChart',
  'BaiduMapChart',
  'HoneycombChart',
  'OrganizationChart',
  'AutonaviMapChart',
  'SnowFlakeChart',
  'TimelineChart',
  'MilestoneChart'
]

// 图表核心对象，按需引入图表 class 给 CoreChart 渲染，打包容量较小
export default class CoreChart extends BaseChart {
  constructor() {
    super()
    // 图表echarts实例
    this.echartsIns = null
    // 图表icharts实例
    this.ichartsIns = null
    // 图表echarts配置项
    this.eChartOption = null
    // 图表icharts配置项
    this.iChartOption = null
    // 图表所在容器
    this.dom = null
    // 图表类型
    this.chartClass = null
    // 图表依赖的三方插件
    this.plugins = {}
    // 图表还没有执行render()方法
    this.hasRender = false
    // 图表渲染完毕的回调
    this.renderCallBack = null
    // 图表resize节流时间
    this.resizeThrottle = 0
    // 图表容器的宽高变化监听器
    this.resizeObserver = null
    // 响应式布局的监听器
    this.mediaScreenObserver = undefined
  }

  // 注册主题
  static registerTheme(name, config) {
    if (!config) {
      tips.error('The second parameter config is required')
      return
    }
    Theme.set(name, config)
  }

  // 注册配置
  static registerConfig(name, config) {
    if (!config) {
      tips.error('The second parameter config is required')
      return
    }
    Theme.setConfig(name, config)
  }

  // 设置主题
  static theme(name) {
    Theme.setDefaultTheme(name)
  }

  static resetThemeCongfig() {
    Theme.resetThemeCongfig()
  }

  // 开启响应式布局（类媒体查询效果）
  mediaScreen(dom, screenOption) {
    this.mediaScreenObserver = new MediaScreen(dom, screenOption, (option) => {
      this.setSimpleOption(this.chartClass, option, this.plugins, false)
      this.render()
    })
  }

  // 初始化echarts，并同时监听容器和窗口的大小变化
  init(chartDom, initOpts) {
    const defaultInit = {
      domResize: true,
      windowResize: true,
      resizeThrottle: this.resizeThrottle
    }
    initOpts = merge(defaultInit, initOpts)
    this.dom = chartDom
    this.echartsIns = echarts.init(chartDom, {}, initOpts)
    // resize节流函数
    this.throttleResize =
      initOpts.resizeThrottle === 0
        ? this.setResize.bind(this)
        : throttle(initOpts.resizeThrottle, this.setResize.bind(this))
    // 容器大小变化监听
    initOpts.domResize && this.setResizeObserver()
    // 页面大小变化监听
    initOpts.windowResize && window.addEventListener('resize', this.throttleResize)
  }

  setResizeObserver() {
    this.resizeObserver = new ResizeObserver((entries) => {
      window.requestAnimationFrame(() => {
        this.throttleResize()
      })
    })
    this.resizeObserver.observe(this.dom)
  }

  // 图表宽高自适应
  setResize() {
    this.mediaScreenObserver && this.mediaScreenObserver.observe()
    this.echartsIns && this.echartsIns.resize && this.echartsIns.resize({ width: 'auto' })
    this.ichartsIns && this.ichartsIns.resize && this.ichartsIns.resize()
  }

  // 传入简化后的icharts-option
  setSimpleOption(ChartClass, iChartOption, plugins = {}, isInit = true) {
    if (isInit) {
      Theme.setDefaultTheme(iChartOption.theme)
      this.mediaScreenObserver && this.mediaScreenObserver.setInitOption(iChartOption)
    }
    if (iChartOption.readScreen) {
      readScreen(this.dom, iChartOption.readScreen)
    }
    if (this.isSelfChart(ChartClass)) {
      this.redirectSelfChart(ChartClass, iChartOption)
      return
    }
    this.plugins = plugins
    this.chartClass = ChartClass
    this.iChartOption = iChartOption
    this.ichartsIns = new ChartClass(iChartOption, this.echartsIns, this.plugins)
    this.eChartOption = this.ichartsIns.getOption()
    this.iChartOption.axistip && axistip(this.dom, this.echartsIns, this.eChartOption)
    mergeExtend(this.iChartOption, this.eChartOption)
  }

  // 若自研图表，走自研图表路径，并更改this指向
  redirectSelfChart(SelfChart, option, plugins) {
    const stateDom = this.dom.getElementsByClassName('huicharts-state-container')[0]
    this.uninstall()
    this.dom.innerHTML = ''
    const instance = new SelfChart()
    instance.init(this.dom)
    instance.setSimpleOption(SelfChart, option, plugins)
    instance.renderCallBack = this.renderCallBack
    if (stateDom) {
      this.dom.appendChild(stateDom)
    }
    Object.setPrototypeOf(this, instance)
    for (const key in this) {
      if (Object.hasOwnProperty.call(this, key)) {
        delete this[key]
      }
    }
  }

  // 判断是否为若自研图表
  isSelfChart(chartClass) {
    return SELF_CHART.includes(chartClass.name)
  }

  // 渲染图表
  render(option) {
    // 已经开始渲染
    this.hasRender = true
    // 第一次渲染
    this.setOption(this.eChartOption, option)
    // 第二次渲染
    this.setOptionAgain(this.eChartOption)
    // 图表渲染完成时回调
    this.renderCallBack && this.renderCallBack(this.echartsIns)
  }

  // 第一次渲染: 调用echarts原生的setOption
  setOption(eChartOption, option) {
    option = merge({ notMerge: true }, option)
    eChartOption && this.echartsIns.setOption(eChartOption, option)
  }

  // 第二次渲染: 有些图表需要根据第一次渲染出来的结果进行二次计算
  setOptionAgain() {
    if (this.ichartsIns && this.ichartsIns.updateOptionAgain) {
      const YAxiMax = this.getYAxisMaxValue(0)
      const YAxiMin = this.getYAxisMinValue(0)
      // 根据网格重新计算option
      this.ichartsIns.updateOptionAgain(YAxiMax, YAxiMin)
      // 再次渲染
      this.setOption(this.eChartOption)
    }
  }

  // 图表刷新，包括刷新配置和数据
  refresh(iChartOption) {
    this.iChartOption = iChartOption
    this.setSimpleOption(this.chartClass, iChartOption, this.plugins)
    this.render()
  }

  // 图表刷新，仅刷新数据
  refreshData(data) {
    this.iChartOption.data = data
    this.refresh(this.iChartOption)
  }

  // 图表渲染完成时回调
  onRenderReady(callback) {
    this.renderCallBack = callback
  }

  // 给echarts单独绑定事件
  on(...rest) {
    this.echartsIns && this.echartsIns.on(...rest)
  }

  // 给echarts单独解绑事件
  off(...rest) {
    this.echartsIns && this.echartsIns.off(...rest)
  }

  // 给echarts实例绑定事件
  bindEvents(events) {
    if (events && events.length !== 0) {
      events.forEach((item) => {
        if (item.query) {
          this.off(item.eventName, item.handler)
          this.on(item.eventName, item.query, item.handler)
        } else {
          this.off(item.eventName, item.handler)
          this.on(item.eventName, item.handler)
        }
      })
    }
  }

  // 批量给echarts实例解绑事件
  unbindEvents(events) {
    if (events && events.length !== 0) {
      events.forEach((item) => {
        if (item.handler) {
          this.off(item.eventName, item.handler)
        } else {
          this.off(item.eventName)
        }
      })
    }
  }

  // 卸载图表时,
  uninstall() {
    // 卸载window resize监听功能
    window.removeEventListener('resize', this.throttleResize)
    // 卸载container容器变化监听
    if (this.resizeObserver) {
      this.resizeObserver.unobserve(this.dom)
      this.resizeObserver.disconnect()
      this.resizeObserver = null
    }
    // 销毁ECharts实例
    if (this.echartsIns && !this.echartsIns.isDisposed()) {
      this.echartsIns.dispose()
    }
    this.echartsIns = null
  }

  // 获取到ECharts实例
  getEchartsInstance() {
    return this.echartsIns
  }

  // 直接传入ECharts的原生配置项
  setEchartsOption(option) {
    option && (this.eChartOption = option)
  }

  // 获取到ECharts配置项
  getEchartsOption() {
    return this.eChartOption
  }

  /**
   * 图表渲染完毕后，获得y刻度的最大值
   * @param {index为yAxis数组下标} index
   * _extent是一个数组，_extent[0]为该轴上最小值，_extent[1]为该轴上最大值
   */
  getYAxisMaxValue(index) {
    return this.echartsIns.getModel().getComponent('yAxis', index)?.axis.scale._extent[1]
  }

  /**
   * 图表渲染完毕后，获得y轴刻度的最小值
   * @param {index为yAxis数组下标} index
   * _extent是一个数组，_extent[0]为该轴上最小值，_extent[1]为该轴上最大值
   */
  getYAxisMinValue(index) {
    return this.echartsIns.getModel().getComponent('yAxis', index)?.axis.scale._extent[0]
  }
}
