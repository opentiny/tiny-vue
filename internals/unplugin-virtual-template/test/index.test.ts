import { describe, expect, it } from 'vitest'
import { transformVirtualTemplate, transformVirtualTemplateForBuild } from '../src/core/index'

describe('index', () => {
  describe('transformVirtualTemplate', () => {
    it('pc|mobile|mobile-first', () => {
      expect(
        transformVirtualTemplate("import template from 'virtual:template?pc|mobile|mobile-first'")
      ).toMatchSnapshot()
    })

    it('pc|mobile=pc|mobile-first', () => {
      expect(
        transformVirtualTemplate("import template from 'virtual:template?pc|mobile=pc|mobile-first'")
      ).toMatchSnapshot()
    })

    it('pc|mobile|mobile-first=pc', () => {
      expect(
        transformVirtualTemplate("import template from 'virtual:template?pc|mobile|mobile-first=pc'")
      ).toMatchSnapshot()
    })
  })

  describe('transformVirtualTemplateForBuild', () => {
    it('pc|mobile|mobile-first', () => {
      expect(
        transformVirtualTemplateForBuild("import template from 'virtual:template?pc|mobile|mobile-first'")
      ).toMatchSnapshot()
    })

    it('pc|mobile=pc|mobile-first', () => {
      expect(
        transformVirtualTemplateForBuild("import template from 'virtual:template?pc|mobile=pc|mobile-first'")
      ).toMatchSnapshot()
    })

    it('pc|mobile|mobile-first=pc', () => {
      expect(
        transformVirtualTemplateForBuild("import template from 'virtual:template?pc|mobile|mobile-first=pc'")
      ).toMatchSnapshot()
    })
  })
})
