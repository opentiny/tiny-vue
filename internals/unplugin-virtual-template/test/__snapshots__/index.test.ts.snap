// Vitest Snapshot v1

exports[`index > transformVirtualTemplate > pc|mobile=pc|mobile-first 1`] = `
"
const template = (mode) => {
  let result

  // #v-ifndef VITE_TINY_MODE
  result = import.meta.glob('./{pc,mobile,mobile-first}.vue', { eager: true })
  // #v-endif

    
  // #v-ifdef VITE_TINY_MODE=pc
  result = import.meta.glob('./pc.vue', { eager: true })
  // #v-endif
          
  // #v-ifdef VITE_TINY_MODE=mobile
  result = import.meta.glob('./pc.vue', { eager: true })
  // #v-endif
          
  // #v-ifdef VITE_TINY_MODE=mobile-first
  result = import.meta.glob('./mobile-first.vue', { eager: true })
  // #v-endif
        

  const finalMode = {\\"pc\\":\\"pc\\",\\"mobile\\":\\"pc\\",\\"mobile-first\\":\\"mobile-first\\"}[mode?.value || mode]

  return result[\`./\${finalMode}.vue\`].default
}
  "
`;

exports[`index > transformVirtualTemplate > pc|mobile|mobile-first 1`] = `
"
const template = (mode) => {
  let result

  // #v-ifndef VITE_TINY_MODE
  result = import.meta.glob('./{pc,mobile,mobile-first}.vue', { eager: true })
  // #v-endif

    
  // #v-ifdef VITE_TINY_MODE=pc
  result = import.meta.glob('./pc.vue', { eager: true })
  // #v-endif
          
  // #v-ifdef VITE_TINY_MODE=mobile
  result = import.meta.glob('./mobile.vue', { eager: true })
  // #v-endif
          
  // #v-ifdef VITE_TINY_MODE=mobile-first
  result = import.meta.glob('./mobile-first.vue', { eager: true })
  // #v-endif
        

  const finalMode = {\\"pc\\":\\"pc\\",\\"mobile\\":\\"mobile\\",\\"mobile-first\\":\\"mobile-first\\"}[mode?.value || mode]

  return result[\`./\${finalMode}.vue\`].default
}
  "
`;

exports[`index > transformVirtualTemplate > pc|mobile|mobile-first=pc 1`] = `
"
const template = (mode) => {
  let result

  // #v-ifndef VITE_TINY_MODE
  result = import.meta.glob('./{pc,mobile,mobile-first}.vue', { eager: true })
  // #v-endif

    
  // #v-ifdef VITE_TINY_MODE=pc
  result = import.meta.glob('./pc.vue', { eager: true })
  // #v-endif
          
  // #v-ifdef VITE_TINY_MODE=mobile
  result = import.meta.glob('./mobile.vue', { eager: true })
  // #v-endif
          
  // #v-ifdef VITE_TINY_MODE=mobile-first
  result = import.meta.glob('./pc.vue', { eager: true })
  // #v-endif
        

  const finalMode = {\\"pc\\":\\"pc\\",\\"mobile\\":\\"mobile\\",\\"mobile-first\\":\\"pc\\"}[mode?.value || mode]

  return result[\`./\${finalMode}.vue\`].default
}
  "
`;

exports[`index > transformVirtualTemplateForBuild > pc|mobile=pc|mobile-first 1`] = `
"
import PcTemplate from './pc.vue'
import MobileTemplate from './pc.vue'
import MobileFirstTemplate from './mobile-first.vue'

const template = (mode) => {
  if ('mobile' === (process.env.TINY_MODE || mode)) return MobileTemplate
  else if ('mobile-first' === (process.env.TINY_MODE || mode)) return MobileFirstTemplate
  else return PcTemplate
}
  "
`;

exports[`index > transformVirtualTemplateForBuild > pc|mobile|mobile-first 1`] = `
"
import PcTemplate from './pc.vue'
import MobileTemplate from './mobile.vue'
import MobileFirstTemplate from './mobile-first.vue'

const template = (mode) => {
  if ('mobile' === (process.env.TINY_MODE || mode)) return MobileTemplate
  else if ('mobile-first' === (process.env.TINY_MODE || mode)) return MobileFirstTemplate
  else return PcTemplate
}
  "
`;

exports[`index > transformVirtualTemplateForBuild > pc|mobile|mobile-first=pc 1`] = `
"
import PcTemplate from './pc.vue'
import MobileTemplate from './mobile.vue'
import MobileFirstTemplate from './pc.vue'

const template = (mode) => {
  if ('mobile' === (process.env.TINY_MODE || mode)) return MobileTemplate
  else if ('mobile-first' === (process.env.TINY_MODE || mode)) return MobileFirstTemplate
  else return PcTemplate
}
  "
`;
